### PR TITLE
Network: dns lookup can return an IPv6 address

### PR DESCRIPTION
**Pull request by joseluismarti**
## Description
The current `gethostbyname` function can only return IPv4 addresses for the name parameter. 

The `getaddrinfo` function and associated `addrinfo` structure should be used.

Both functions can return one or several addresses, sorted according to predefined rules, the first one is used.

## How has this been tested?
Tested on Linux that name resolution works correctly on an IPv4-only stack interface and also on a dual-stack network interface.

Also check on Linux that it performs name resolution and connects when the target host has only IPv6 addresses, and verified that IPv6 addresses are added to the cache.

I've verified that the name resolution also works as usual on Android.

## Types of change
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/network/DNSNameCache.cpp [hunk 1]
```diff
@@ -58,13 +59,17 @@ bool CDNSNameCache::Lookup(const std::string& strHostName, std::string& strIpAdd
     return true;
 
   // perform dns lookup
-  struct hostent *host = gethostbyname(strHostName.c_str());
-  if (host && host->h_addr_list[0])
+  addrinfo hints;
```
Comment (thexai) [inline xbmc/network/DNSNameCache.cpp:62]:
```suggestion
  addrinfo hints{};
```

And remove `memset`

File: xbmc/network/DNSNameCache.cpp [hunk 2]
```diff
@@ -58,13 +59,17 @@ bool CDNSNameCache::Lookup(const std::string& strHostName, std::string& strIpAdd
...
+  addrinfo* res;
+
+  memset(&hints, 0, sizeof(hints));
+  hints.ai_family = AF_UNSPEC;
+  hints.ai_socktype = SOCK_STREAM;
+  hints.ai_flags |= AI_CANONNAME;
+
+  if (getaddrinfo(strHostName.c_str(), nullptr, &hints, &res) == 0)
   {
-    strIpAddress = StringUtils::Format("{}.{}.{}.{}", (unsigned char)host->h_addr_list[0][0],
```
Comment (thexai) [inline xbmc/network/DNSNameCache.cpp:64]: `#include "utils/StringUtils.h"` seems can be removed as not used in any other place of this file

File: xbmc/network/DNSNameCache.cpp (hunk 1 above)
Comment (joseluismarti) [reply xbmc/network/DNSNameCache.cpp]: Ok, thanks

File: xbmc/network/DNSNameCache.cpp (hunk 2 above)
Comment (joseluismarti) [reply xbmc/network/DNSNameCache.cpp:64]: I didn't realise, thanks

**Review by thexai — APPROVED** [review]

File: xbmc/network/DNSNameCache.cpp [hunk 3]
```diff
@@ -58,13 +58,16 @@ bool CDNSNameCache::Lookup(const std::string& strHostName, std::string& strIpAdd
...
-                                       (unsigned char)host->h_addr_list[0][1],
-                                       (unsigned char)host->h_addr_list[0][2],
-                                       (unsigned char)host->h_addr_list[0][3]);
+    strIpAddress = CNetworkBase::GetIpStr(res->ai_addr);
```
Comment (neo1973) [inline xbmc/network/DNSNameCache.cpp:70]: Missing resource cleanup:
```suggestion
    strIpAddress = CNetworkBase::GetIpStr(res->ai_addr);
    freeaddrinfo(res);
```

**Review by neo1973 — CHANGES_REQUESTED** [review]

File: xbmc/network/DNSNameCache.cpp (hunk 3 above)
Comment (joseluismarti) [reply xbmc/network/DNSNameCache.cpp:70]: You're right, thanks

**Review by neo1973 — APPROVED** [review]

**Pull request merged**